File: engine/src/elements/Align.ts
import Element from '../Element'
import BoxConstraints from '../BoxConstraints'
import Size from '../Size'
import Position from '../Position'

class Alignment {
	value: number
	isRelative: boolean

	constructor(value: number, isRelative: boolean) {
		this.value = value
		this.isRelative = isRelative
	}

	calc(total: number) {
		return this.isRelative ? this.value * total : this.value
	}
}

interface AlignProps {
	child: Element<any>,
	top?: Alignment,
	left?: Alignment,
	bottom?: Alignment,
	right?: Alignment
}

class Align extends Element<AlignProps> {
	sizedByParent = true

	layout(constraints: BoxConstraints): Size {
		const { top, left, bottom, right, child } = this.props
		const horizAlign = (left || right).calc(constraints.maxWidth)
		const vertAlign = (top || bottom).calc(constraints.maxHeight)
		const childConstraints = new BoxConstraints({
			minWidth: 0,
			maxWidth: constraints.maxWidth - horizAlign,
			minHeight: 0,
			maxHeight: constraints.maxHeight - vertAlign
		})
		const size = this.document.layoutElement(child, childConstraints)
		child.size = size
		child.relPosition = new Position({
			left: left ? horizAlign : 0,
			top: top ? vertAlign : 0
		})
		return new Size({
			width: constraints.maxWidth,
			height: constraints.maxHeight
		})
	}

	paint() {
		this.document.paintElement(this.props.child)
	}

	static px(val: number) {
		return new Alignment(val, false)
	}

	static rel(val: number) {
		return new Alignment(val, true)
	}
}

export default Align

File: core/libs/ui/src/elements/align.cpp
#include "elements/align.hpp"

namespace aardvark {

AlignElement::AlignElement(
    std::shared_ptr<Element> child,
    Alignment insets,
    bool adjust_child_size,
    bool is_repaint_boundary)
    : SingleChildElement(
          std::move(child),
          is_repaint_boundary,
          /* size_depends_on_parent */ true),
      adjust_child_size(adjust_child_size),
      insets(insets){};

float AlignElement::get_intrinsic_height(float width) {
    return insets.calc_vert(0) +
           child->get_intrinsic_height(width - insets.calc_horiz(0));
}

float AlignElement::get_intrinsic_width(float height) {
    return insets.calc_horiz(0) +
           child->get_intrinsic_width(height - insets.calc_horiz(0));
}

Size AlignElement::layout(BoxConstraints constraints) {
    auto left = insets.left.calc(constraints.max_width);
    auto top = insets.top.calc(constraints.max_height);
    auto right = insets.right.calc(constraints.max_width);
    auto bottom = insets.bottom.calc(constraints.max_height);
    auto horiz = left + right;
    auto vert = top + bottom;

    auto child_constraints = BoxConstraints{
        0,  // min_width
        adjust_child_size ? (constraints.max_width - horiz)
                          : constraints.max_width,  // max_width
        0,                                          // min_height
        adjust_child_size ? (constraints.max_height - vert)
                          : constraints.max_height,  // max_height
    };
    auto size = document->layout_element(child.get(), child_constraints);
    child->size = size;
    child->rel_position = Position{
        insets.left.is_none() ? (constraints.max_width - horiz - size.width)
                              : left,  // left
        insets.top.is_none() ? (constraints.max_height - vert - size.height)
                             : top  // top
    };
    return constraints.max_size();
};

void AlignElement::paint(bool is_changed) {
    document->paint_element(child.get());
};

}  // namespace aardvark
